src/main/java/org/broadinstitute/hellbender/tools/copynumber/CreateBinningIntervals.java

```diff
@@ -0,0 +1,122 @@
+
+    @Argument(
+            doc = "width of the bins",
+            fullName = WIDTH_OF_BINS_LONG_NAME,
+            shortName = WIDTH_OF_BINS_SHORT_NAME,
+            optional = true,
+            minValue = 1
+    )
+    private int widthOfBins = 1;
+
+    @Argument(
+            doc = "width of the padding regions",
+            fullName = PADDING_LONG_NAME,
+            shortName = PADDING_SHORT_NAME,
+            optional = true,
+            minValue = 0
+    )
+    private int padding = 0;
+
+    @Argument(
+            doc = "output file",
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            optional = false
+    )
+    public File outputFile;
+
+
+
+    @Override
+    public void onTraversalStart() {
+        createBins();
+    }
+
+    /**
+     * Generates binning coverage in the intervals given by the user.
+     * The width of bins, the intervals and the output file's path are given by the user.
+     */
+    public void createBins() {
+        // check if the output directory exists
+        if (!outputFile.exists() && !outputFile.mkdir()) {
+            throw new RuntimeException("Unable to create file: " + outputFile.getAbsolutePath());
+        }
+
+        // check if the bin widths are set appropriately
+        if(widthOfBins <= 0) {
+            throw new IllegalArgumentException("Width of bins " + Integer.toString(widthOfBins) + " should be >= 0.");
+        }
+
+        // get the sequence dictionary
+        final SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+        final List<SimpleInterval> intervals = hasIntervals() ? intervalArgumentCollection.getIntervals(sequenceDictionary)
+               : IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
+
+        // create an IntervalList by copying all elements of 'intervals' into it
+        IntervalList intervalList = new IntervalList(sequenceDictionary);
+        intervals.stream().map(si -> new Interval(si.getContig(), si.getStart(), si.getEnd())).forEach(intervalList::add);
+
+        // sort intervals according to their coordinates and unique them (i.e. delete duplicates)
+        intervalList.uniqued();
+
+        // pad all elements of intervalList
+        intervalList = intervalList.padded(padding,padding);
+
+        // merge those that intersect after padding
+        intervalList = IntervalList.intersection(intervalList, intervalList);
+
+        // break the intervals up to bins -- the last bin in each interval can be shorter than the others
+        IntervalList bins = new IntervalList(sequenceDictionary);
+        int bin_start, bin_end;
+        Interval new_bin;
+        for(Interval in : intervalList) {
+            bin_start = in.getStart();
+            bin_end = Math.min(bin_start + widthOfBins - 1, in.getEnd());
+            while(bin_start < in.getEnd()) {
```

Comment:
This might work better as a for loop:
```java
for(int binStart = in.getStart(); binStart < in.getEnd(); binStart += widthOfBins) {
   final int binEnd = Math.min(bin_start + widthOfBins - 1, in.getEnd());
   bins.add(new Interval(in.getContig(), bin_start, bin_end));
}
```

Reply:
It is a good practice to use `FastMath` instead of `Math`. The former is a drop-in replacement for the latter (i.e. with identical method signatures and semantics), though, with additional optimizations. Simple functions such as `min` and `max` have identical implementations. However, the implementation of more complicated functions such as `cos` (!) is highly optimized. Verbatim from its javadoc: `FastMath speed is achieved by relying heavily on optimizing compilers to native code present in many JVMs today and use of large tables. The larger tables are lazily initialised on first use, so that the setup time does not penalise methods that don't need them.`

Reply:
Thanks for all your comments. I used David's for loop with Mehrtash's FastMath and Sam's <= condition :)